### ISSUE TITLE
Reflection: cannot reduce type because variable is erased

### ISSUE THREAD

**Issue by nad** [agda/agda#6124]
A follow-up to issue #5734:
```agda
open import Agda.Builtin.List
open import Agda.Builtin.Reflection
open import Agda.Builtin.Unit

id : (@0 A : Set) → A → A
id _ x = x

macro

  @0 Unit : Term → TC ⊤
  Unit goal =
    bindTC (inferType (def (quote id) [])) λ t →
    bindTC (reduce t) λ _ →
    unify goal (def (quote ⊤) [])

_ : Set
_ = Unit
```
```
Variable A is declared erased, so it cannot be used here
when checking that the expression A has type _16
```
The reflection machinery cannot reduce the type of `id` because it thinks that the erased variable `A` is used in a non-erased context. However, it is fine to use `A` in this way *in a type*.

This particular problem can be fixed by adding `workOnTypes` to line 779:

https://github.com/agda/agda/blob/1aa7b2e413f0cc68b1d737ac52e0d34a1fbe85f3/src/full/Agda/TypeChecking/Unquote.hs#L776-L787

@jespercockx [wrote](https://github.com/agda/agda/issues/5734#issuecomment-1013880272) the following:
> The only time you get an error "early" in the macro is when you use one of the TC primitives that take a term as input, but there it is really necessary to check that the input is well-formed because otherwise it could violate internal invariants of Agda.

Would any internal invariants be broken by the use of `workOnTypes`?

Should every call to `inferExpr` in `Agda.TypeChecking.Unquote` be wrapped in `workOnTypes`?

**Comment by nad** [comment]
The use of `workOnTypes` broke one of my tactics. The problem seems to be that `workOnTypes` changes how Agda reduces things:

https://github.com/agda/agda/blob/1aa7b2e413f0cc68b1d737ac52e0d34a1fbe85f3/src/full/Agda/TypeChecking/Irrelevance.hs#L117-L126

I used the following code instead, and that seems to work better, but I don't know if this fix breaks anything:
```haskell
      (v, t) <- applyQuantityToContext zeroQuantity $
                locallyReduceAllDefs $ inferExpr =<< toAbstract_ v
```

**Comment by WhatisRT** [comment]
I've ran into this issue as well while trying out 2.6.3 with my code. Since some constructors (like those for sums and products) now automatically have some of their arguments erased, it's basically impossible to do anything with their types in the reflection mechanism anymore. AFAICT there's no way to disable this behaviour, so 2.6.3 is basically completely broken for me 😢

**Comment by jespercockx** [comment]
Could you give a (preferably small) example of what exactly is broken? If there's a clear regression wrt 2.6.2, I believe we should prioritize this to still get a fix into the release of 2.6.3.

**Comment by WhatisRT** [comment]
This is an example:
```agda
macro
  testM : Term → TC ⊤
  testM hole = getType (quote _,_) >>= unify hole

test : Set
test = testM
```

**Comment by jespercockx** [comment]
I think the fix that @nad proposes (adding `applyQuantityToContext zeroQuantity`) sounds reasonable for the TC primitives that just return something of type `Term` without otherwise modifying the TC state (`checkType`, `inferType`, `reduce`, and `normalise`) However, for operations that return an agda term (`unquoteTC`) or modify the TC state (`unify`) this is not a sound approach since it would allow ill-typed terms to seep into actual internal agda syntax. So it's not immediately obvious to me how to fix this.

**Comment by jespercockx** [comment]
Actually, Agda does modality-check meta solutions before instantiating them (see https://github.com/agda/agda/blob/master/src/full/Agda/TypeChecking/MetaVars/Occurs.hs#L166 and https://github.com/agda/agda/blob/master/src/full/Agda/TypeChecking/MetaVars/Occurs.hs#L196 and https://github.com/agda/agda/blob/master/src/full/Agda/TypeChecking/MetaVars/Occurs.hs#L231) so calling `unify` in erased mode should also not be a problem. That just leaves `unquoteTC`.

**Comment by WhatisRT** [comment]
A quick search shows that I don't use `unquoteTC` anywhere, so I guess a partial fix would make me happy for now.

**Comment by jespercockx** [comment]
Actually, after thinking about `unquoteTC` a bit more I don't believe that anything bad could really happen here: the only thing that `unquoteTC` would allow you to do (if it were run in erased mode) is to get access to something erased inside the TC monad, e.g. as follows:
```
postulate @0 X : Set

macro
  testM : Term → TC ⊤
  testM hole = do
    X' ← unquoteTC {A = Set} (def (quote X) [])
    -- have fun with `X'` here
    returnTC _
```
However, there is no way to smuggle the `X'` out of the `TC` monad. 

There is also no way (that I can see) that this can be exploited by doing something bad *within* the TC monad, since we do not get an equality `X' = X`. Here is my best attempt at tricking Agda into believing so:
```agda
postulate @0 X : Set

@0 tricky : Σ Set (X ≡_)
tricky = X , refl

macro
  testM : Term → TC ⊤
  testM hole = do
    (X' , refl) ← unquoteTC {A = Σ Set (X ≡_)} (def (quote tricky) [])
    -- Can we have `X' = X` definitionally here?
    returnTC _
```
This fails because `unquoteTC` requires the type argument `A` to be *not* erased:
```
Identifier X is declared erased, so it cannot be used here
when checking that the expression X has type _A_16
```
This makes sense since `unquoteTC` really does make use of the `A` argument, so the fact that it is not erased is essential. 

In other words, allowing `unquoteTC` to be run on erased things is equivalent to adding a primitive `unerase : @0 A -> TC A`, which is not harmful since `TC` computations do not have any computational behaviour in internal syntax, hence we might as well assume that `TC A = ⊤`. So I think we can go ahead and do the fix proposed by @nad. I'll prepare a PR.

**Comment by jespercockx** [comment]
I've tried the approach I outlined before, adding `applyQuantityToContext zeroQuantity` to the appropriate places in the execution of TC operations. However, I've run into one more snag: with this change **all** metavariables that are created whilst unquoting a reflected term (e.g. when unquoting `unknown`) are considered to be in an erased context. In most cases this doesn't matter because erased metas can later be promoted to non-erased status, but when using the elaborate-and-give command it does cause issues:
```agda
open import Agda.Builtin.List
open import Agda.Builtin.Nat
open import Agda.Builtin.Reflection renaming (bindTC to _>>=_)
open import Agda.Builtin.Unit

vArg : {A : Set} → A → Arg A
vArg = arg (arg-info visible (modality relevant quantity-ω))

macro
  mcMacro : Term → TC ⊤
  mcMacro hole =  unify hole (def (quote _+_) (vArg unknown ∷ vArg unknown ∷ []))

test : Nat
test = {!mcMacro!}
```
Running `C-c C-m` here (with my changes) doesn't instantiate the hole at all, but instead reports an unsolved constraint:
```
———— Errors ————————————————————————————————————————————————
Failed to solve the following constraints:
  ?1 = _15 + _16 : Nat (blocked on any(_12, _15))
```
The reason is that the hole only accepts non-erased solutions, but both arguments to `_+_` are erased metas. Moreover, the second meta is in a flexible context so the promotion to non-erased status doesn't happen. I might need a bit more time to come up with a better solution to this :/

Edit: here is my attempt, which produces a regression in `interaction/ElaborateAndGive.agda`: https://github.com/jespercockx/agda/tree/issue6124-attempt1

**Comment by jespercockx** [comment]
Coming back to the original issue, I guess I still don't really understand why it is not allowed to freely use the argument of an erased function type in its codomain, i.e. why is the following not allowed:
```agda
postulate Foo : Nat → Set
foo : Set
foo = (@0 n : Nat) → Foo n
```
If we could just allow this code, then there would be no need for all the extra complexity that comes with allowing erased things inside TC code.

**Comment by jespercockx** [comment]
Actually apparently I made this change before and it was undone by @Saizan : https://github.com/agda/agda/commit/78b956beb5045480326a906a5949c185c5e781ab 

I don't really understand why it was considered to be "ad hoc" though, it feels to me like it is the right thing to do. Perhaps the modality experts @JorisCeulemans @anuyts, @flupe and @jpoiret want to weigh in on this issue?

**Comment by WhatisRT** [comment]
I'm not very familiar with how erasure works/is intended to work in Agda but assuming it's similar to Cedille/CDLE, I think you're right that the example you give should work. The example given in #3853 is odd, since one of the variables is bound with a pi and the other one with a lambda. The restrictions for erasure should probably just apply to a lambda bound variable.

**Comment by nad** [comment]
I [asked](https://github.com/agda/agda/issues/4784#issuecomment-1287803484) a similar question for the following type:
```agda
record R : Set₁ where
  field
    @0 A : Set
    x    : A
```
I received an answer from @Saizan via some other channel. He noted that the typing rules for erasure have been designed to ensure that definitional equality preserves erasure, and that this property is used in a proof that shows that it is OK to erase transports.

**Comment by jespercockx** [comment]
> I received an answer from @Saizan via some other channel. He noted that the typing rules for erasure have been designed to ensure that definitional equality preserves erasure, and that this property is used in a proof that shows that it is OK to erase transports.

I'm not sure how to interpret "it is OK to erase transports". Do you mean this type should or should not be allowed?

**Comment by anuyts** [comment]
I'm neither familiar with reflection nor with the Agda code base (even after the AIM) but as far as modalities are concerned:
I think the following should be allowed:
```agda
postulate Foo : Nat → Set
foo : Set
foo = (@0 n : Nat) → Foo n
```
In general, I think the typing rules for lambda and pi should be different:
```
Γ, @0 x : A |– b : B
--------------------------
Γ |– λ (@0 x : A) -> b : (@0 x : A) -> B
```
but
```
Γ, @ω x : A |– B : Set
---------------------------------
Γ |– (@0 x : A) -> B : Set
```
This is reconciled by the fact that the judgement
```
Γ |– t : T
```
presupposes
```
@0 \ Γ |– T : Set _
```

I'm not sure what is the meaning of what @nad says @Saizan says, neither "definitional equality preserves erasure" nor "it is OK to erase transports".

**Comment by anuyts** [comment]
@pigworker also seems to agree, see the FUN-rule on p6 of
https://personal.cis.strath.ac.uk/conor.mcbride/PlentyO-CR.pdf
(I'm assuming that absence of a quantity means ω).

**Comment by anuyts** [comment]
Oh actually no, his type judgements have a turnstile annotated with `0` so I suspect that he only derives types when you're in erased mode to begin with.

**Comment by nad** [comment]
> I'm not sure how to interpret "it is OK to erase transports".

The erasure function for Erased Cubical Agda contains the clause `|primTransp _ _ e| = |e|`. This would not work for full Cubical Agda, but Erased Cubical Agda is set up so that this should work.

> Do you mean this type should or should not be allowed?

I don't know if this type is fine or not, and I don't think we should allow it without some kind of supporting theory. (I imagine that the type is fine if `--with-K` is used.)

**Comment by jespercockx** [comment]
> I don't know if this type is fine or not, and I don't think we should allow it without some kind of supporting theory. (I imagine that the type is fine if --with-K is used.)

@anuyts mentioned above that in general we should not use the same mode for variables bound by a lambda or bound by a pi. This seems to be supported by the typing rules in the following paper: https://lmcs.episciences.org/7713/pdf (pages 15 & 16). This does not include cubical type theory, however I believe asking for a fully developed theory of all possible combinations of features before we implement them in Agda is an unrealistic standard (and also one that is definitely not followed by other features of Agda). Instead, I think our criteria should be that we can add a feature if (1) it is consistent with the currently known theoretical developments and (2) does not have any currently known soundness issues when combined with other Agda features.

**Comment by nad** [comment]
> @pigworker also seems to agree, see the FUN-rule on p6 of https://personal.cis.strath.ac.uk/conor.mcbride/PlentyO-CR.pdf (I'm assuming that absence of a quantity means ω).

That paper is not about cubical type theory. Some typing rules for erasure that make sense in plain type theory do not work in cubical type theory.

> I'm not sure what is the meaning of what @nad says @Saizan says, neither "definitional equality preserves erasure" nor "it is OK to erase transports".

I'm not sure exactly what "definitional equality preserves erasure" (which is not a direct quote from what @Saizan wrote) refers to, but in ["Compiling Programs with Erased Univalence"](http://www.cse.chalmers.se/~nad/publications/abel-danielsson-vezzosi-erased-univalence.html) it is stated that if Γ ⊢ t₁ ⇒ t₂ :ᵖ T holds, then Γ ⊢ t₁ :ᵖ T, Γ ⊢ t₂ :ᵖ T and Γ ⊢ t₁ = t₂ : T should all hold (see Section 4.2).

**Comment by jespercockx** [comment]
> I'm not sure exactly what "definitional equality preserves erasure" (which is not a direct quote from what @Saizan wrote) refers to

You are the one who used that phrasing earlier in this issue.

**Comment by nad** [comment]
> You are the one who used that phrasing earlier in this issue.

I just wanted to make clear that I was not quoting but paraphrasing @Saizan.

> This does not include cubical type theory, however I believe asking for a fully developed theory of all possible combinations of features before we implement them in Agda is an unrealistic standard (and also one that is definitely not followed by other features of Agda). Instead, I think our criteria should be that we can add a feature if (1) it is consistent with the currently known theoretical developments and (2) does not have any currently known soundness issues when combined with other Agda features.

In what sense is the feature "consistent with the currently known theoretical developments"? The draft paper ["Compiling Programs with Erased Univalence"](http://www.cse.chalmers.se/~nad/publications/abel-danielsson-vezzosi-erased-univalence.html) uses the currently implemented typing rule. I'm not aware of any other theoretical work on erasure for Erased Cubical Agda.

However, I have no problem with changing the typing rule for the `--with-K` variant of Agda.

**Comment by anuyts** [comment]
Well `|primTransp _ _ e| = |e|` suggests that the property that we need is that *propositionally* equal closed types erase to *definitionally* equal types.

MTT in itself does not support a special treatment of types. (Also note that modes ≠ modalities.) The only *modal* (as opposed to quantitative) papers that I'm aware of that do this are the ones on [ParamDTT](https://anuyts.github.io/files/paramdtt.pdf) and [RelDTT](https://anuyts.github.io/files/paper-reldtt.pdf). The easiest way to understand what's going on, imo, is to imagine that there is an (implicitly inserted) modal function from the syntactically available universe `Set` to a semantic universe `SetSem`:
```agda
El : @par Set -> SetSem
```
so that we can do
```
@par \ Γ |– A : Set
-----------------
Γ |– El A : SetSem
-----------------
Γ |– El A type
-----------------
Γ, x : El A ctx
```
In ParamDTT and RelDTT, `@par` would be the parametricity modality. For erasure, it would be `@0`.

However, I wasn't aware that a theoretical foundation of erased cubical was available. This one clearly uses a type system that has quantity annotations on the right, which is not (obviously) an instance of MTT. (It might be a multimode system over the poset of quantities and the "divides" relation.) So it would make sense to use that paper as the guide here.

**Comment by jespercockx** [comment]
After going through the paper by @Saizan @nad and @andreasabel I got convinced that in the context of cubical type theory it is indeed not desirable to allow the type `(@0 x : Nat) -> P x` as a term of type `Set` if `P : Nat -> Set` but only when `P : @0 Nat -> Set` (see my comment https://github.com/agda/agda/pull/6396#issuecomment-1343006786). However, that leaves us with no good options to fix this issue:
- The status quo is problematic since it does not allow us to do operations on the result of `getType` when that type is an erased function type.
- Making all TC operations work in an erased context is problematic since it means `unknown` is then always unquoted to an erased meta.
- Changing the typing rule for erased dependent pi types is problematic since it breaks the computation rule for --erased-cubical.

One other option I see would be to duplicate the current TC primitives to have separate versions that work on terms and types, but this is something that I'd prefer to avoid if possible.

**Comment by anuyts** [comment]
I think the reflection interface should allow the user to change the quantity annotation on the right, by analogy to `inContext` and `extendContext`.

**Comment by jespercockx** [comment]
> I think the reflection interface should allow the user to change the quantity annotation on the right, by analogy to `inContext` and `extendContext`.

Ah yes, that would only require adding a single new primitive. Perhaps that's the way we should go then. Should the primitive just go into erased mode (like `applyQuantityToContext zeroQuantity` does) or should it do everything that `workOnTypes` does? For reference, `workOnTypes` currenty does the following other things:

- Enable certain reductions in the confluence checker (if `--confluence-check` or `--local-confluence-check` is enabled)
- Go into nonstrict mode (if `--experimental-irrelevance` is enabled)
- Turn off unfolding of inlined functions

I guess none of these things really matter for the reflection API, so we could go either way.

**Comment by anuyts** [comment]
I'm not sure what the non-strict mode does.
Being able to turn off unfolding seems important, so I would suggest to expose the full `workOnTypes` operation, but maybe also the fine-grained one for just setting the RHS quantity.
(But I have almost no experience with reflection.)

**Comment by jespercockx** [comment]
> I'm not sure what the non-strict mode does.

I don't think anyone really does.

**Comment by andreasabel** [comment]
> I'm not sure what the non-strict mode does.

If you want `(@irr x : A) -> B(x)`, then `B : @shirr A -> Set` is required at least (of course, `B : @irr A -> B` is uncontroversially ok).  Family `B` needs to be shape-irrelevant in `x`, otherwise eta-equality at the instances of `B` is not uniform, thus, ill-defined.

**Comment by jpoiret** [comment]
> > I'm not sure what the non-strict mode does.
> 
> If you want `(@irr x : A) -> B(x)`, then `B : @shirr A -> Set` is required at least (of course, `B : @irr A -> B` is uncontroversially ok). Family `B` needs to be shape-irrelevant in `x`, otherwise eta-equality at the instances of `B` is not uniform, thus, ill-defined.

In the polarity PR (https://github.com/agda/agda/pull/6385) @anuyts added a comment https://github.com/agda/agda/pull/6385/commits/fc71eda87697989e39d31aa41f2ee32c907c5759 describing exactly that as well, so maybe a low-hanging fruit would be to rename `NonStrict` to `ShapeIrrelevance` in the code base as well, and document the feature in the docs? Would that suit everyone? I could probably do that after we finish polishing up the polarity PR.

**Comment by jespercockx** [comment]
> In the polarity PR (https://github.com/agda/agda/pull/6385) @anuyts added a comment https://github.com/agda/agda/commit/fc71eda87697989e39d31aa41f2ee32c907c5759 describing exactly that as well, so maybe a low-hanging fruit would be to rename NonStrict to ShapeIrrelevance in the code base as well, and document the feature in the docs? Would that suit everyone? I could probably do that after we finish polishing up the polarity PR.

Sounds good to me, if you actually have an idea of how shape-irrelevance should behave. In the past there has been quite a bit of confusion around its intended semantics, and in particular it was often confused with runtime irrelevance (back before when we had `@0`). So I think the first step would be indeed to write down a clear design of what it is supposed to be and what the typing rules are.

See also https://github.com/agda/agda/issues/1114 and https://github.com/agda/agda/issues/2595 (edit: and https://github.com/agda/agda/issues/2640 and https://github.com/agda/agda/issues/2757)

**Comment by anuyts** [comment]
It seems I've given a pretty good specification of a system with pointwise, continuous, parametric, shape-irrelevant, parametric&shape-irrelevant, and irrelevant modalities in §9.2.2 of my PhD thesis:
https://anuyts.github.io/files/phd.pdf
Shape-irrelevance and irrelevance actually interact via parametricity, so I'm afraid it's hard to cherry-pick this modal system.

**Comment by anuyts** [comment]
[Normalization by Evaluation for Sized Dependent Types](https://dl.acm.org/doi/pdf/10.1145/3110277) does have a system with only irrelevance and shape-irrelevance but you'll miss that shape-irrelevant usages in parametric positions are ultimately irrelevant (`par º shi = irr`).

**Comment by anuyts** [comment]
> > I'm not sure what the non-strict mode does.
> 
> If you want `(@irr x : A) -> B(x)`, then `B : @shirr A -> Set` is required at least (of course, `B : @irr A -> B` is uncontroversially ok). Family `B` needs to be shape-irrelevant in `x`, otherwise eta-equality at the instances of `B` is not uniform, thus, ill-defined.

In what sense is this a "mode", i.e. what does `workOnTypes` do here exactly?

**Comment by WhatisRT** [comment]
For the record, I've tried the branch exposing `workOnTypes` and got my metaprograms to work again.

**Comment by jespercockx** [comment]
The change that marked parameters as erased in the types of constructors was reverted by https://github.com/agda/agda/pull/6405, so this is no longer a regression. Hence I am bumping it to the next release.

**Comment by andreasabel** [comment]
@jespercockx : Plans to fix this for 2.6.4?

**Comment by jespercockx** [comment]
I think this has been solved by #6405 + the introduction of the `workOnTypes` TC primitive, with the remaining problem being described in https://github.com/agda/agda/issues/6400, so I'm closing this one as completed. This has a long discussion though so I might have missed something, if that's the case please create a new issue for it.

**Comment by jespercockx** [comment]
Somehow I was absolutely convinced that #6399 got merged a long time ago, but apparently this is not actually the case and we still do not have a `workOnTypes` TC primitive. I'll see if I can resurrect that PR, since this issue is currently still unsolved as far as I can tell.
